### DISCUSSION TITLE
Trigger rebuild/flush state/notifyListeners in StateNotifier

### DISCUSSION THREAD

**Discussion by Petri-Oosthuizen** [rrousselGit/riverpod#1962] [Q&A]
Is it possible to `notifyListeners` and trigger a rebuild within a single StateNotifier  function?

The following code triggers a rebuild once:
```
class ExampleNotifier extends StateNotifier<ExampleState> {
  Future<void> task() async {
    state = state.copyWith(value: null);
    // some logic which might take a while or complete instantly
    state = state.copyWith(value: true);
  }
}
```

If the logic between the state updates completes instantly the widget is only rebuilt once. If the logic takes a few ms (awaiting an async result) the tree is rebuilt. This is good for performance and what you want 99% of the time but sometimes I would like to flush the state, akin to `notifyListeners`. I mostly want this to reset components be that is forms or animations.

**Comment by rrousselGit** [comment]
No, it's not possible.

Flutter naturally does this. Calling `setState` twice on a widget only rebuilds it once.

**Comment by TimWhiting** [comment]
If you want it to flush the state, you would need to create a method that if it completes instantly it inserts and artificial delay (awaits a Future).

**Comment by Petri-Oosthuizen** [comment]
@TimWhiting I've found that to work but I don't really like the idea of inserting an artificial delay.
